spyder/plugins/variableexplorer.py

```diff
@@ -164,9 +168,7 @@ def add_shellwidget(self, shellwidget):
         shellwidget_id = id(shellwidget)
         if shellwidget_id not in self.shellwidgets:
             self.options_button.setVisible(True)
-            nsb = NamespaceBrowser(self,
-                                   options_button=self.options_button,
-                                   plugin_actions=[self.undock_action])

```

Comment:
Why did you remove the `undock_action` from here?

Reply:
Presently, actions are added to the menu of the `options_button` during the setup of the `NamespaceBrowser` widget. This is a problem because the `options_button` is shared among all the namespace browsers of the variable explorer plugin (which are added to a `StackWidget`). Therefore, the `options_button` ends up with the actions of the last namespace browser added to the plugin and the actions are not refreshed when the current namespace browser changes.

To fix this issue, I'm proposing to make the  variable explorer plugin responsible to add the actions to the `options_button` instead of doing it in the widget . Therefore, the `plugin_actions` argument is kept here only as a convenience when using the  `NamespaceBrowser` widget outside of the `VariableExplorer` plugin. As you can see, the `NamespaceBrowser.plugin_actions attribute` is only use if the `options_button` argument is `None` when instantiating the `NamespaceBrowser` widget. Please see https://github.com/spyder-ide/spyder/pull/7710/files#diff-6b96bad69031242aedccaa716731ff49R217

So, since we do not override the actions in the widget anymore, we do not need to pass the `undock_action` action when instantiating the `NamespaceBrowser` in the `VariableExplorer` plugin, because this is already added by default to the menu of the `options_button` when the plugin is initialize. See https://github.com/spyder-ide/spyder/blob/master/spyder/api/plugins.py#L102.

Besides, each time the current namespace browser change, the actions of the `options_buttons` are refreshed by the variable explorer plugin using the [refresh_actions](https://github.com/spyder-ide/spyder/blob/2208ff88834bc575383c61d2bf8096fb52ad5642/spyder/api/plugins.py#L209) method of the plugin api, which automatically append the `undock_action` to the plugin actions.

Sorry for the long answer, I hope my explanations were ok.

Reply:
Ok, thanks for the really nice explanation!